docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1844,16 +1855,94 @@ entry value. When providing serializable enum values through `P4Data`, one can
 either use the enum entry's name (`enum` human-readable string field) or its
 assigned value (`enum_value` bytestring field).
 
+### User-defined types { #sec-user-defined-types}
+
+P4~16~ enables programmers to introduce new types [@P4NewTypes]. While similar
+to `typedef`, this mechanism introduces in fact a new type, which is not
+strictly synonym with the original type. It is important to preserve this
+distinction in the P4Info message, in particular for the purposes of
+[translation](#sec-psa-metadata-translation). When introducing a new type, the
+declaration can be annotated with `@p4runtime_translation` to indicate that the
+type exposed to the P4Runtime client is different from the original P4 type. One
+important use-case if for [port numbers](#sec-translation-of-port-numbers),
+whose underlying dataplane representation may vary on different targets, but for
+which it may be convenient to present a unified representation and numbering
+scheme to the control-plane. **The `@p4runtime_translation` annotation can only
+be used if the underlying P4 built-in type is a fixed-width unsigned bitstring
+type (`bit<W>`)** and the type exposed to the control-plane will also be a
+fixed-width unsigned bitstring, with a potentially different bitwidth. It takes
+two parameters: a *URI* (Uniform Resource Identifier) which uniquely identifies
+the translation being performed on entities of the new type to the P4Runtime
+server and the *bitwidth* of the bitstring type exposed to the control-plane. It
+is recommended that the URI includes at least the P4 architecture name and the
+type name.
+
+User-defined types are specified using the `P4NewTypeSpec` message, which has
+the following fields:
+
+* `original_type`, which specifies the underlying type used in the P4 `type`
+  declaration, using the generic `P4DataTypeSpec` message. The `original_type`
+  can correspond to a P4~16~ built-in type or to another user-defined type.
+
+* `translation` of type `P4NewTypeTranslation`, which is set if and only if the
+  P4 type declaration was annotated with `@p4runtime_translation`. The
+  `P4NewTypeTranslation` message itself has two fields - `uri` and
+  `sdn_bitwidth`, which map to the two input parameters to the annotation.
+
+* `annotations`, a repeated field of strings, each one representing a P4
+  annotation associated to the type declaration.
+
+For example, an architecture - in this case PSA - may introduce a new type for
+port numbers:
+~ Begin P4Example
+@p4runtime_translation("p4.org/psa/v1/PortId_t", 32)
```

Comment:
I'm not deeply familiar with URIs, but at least the Wikipedia page makes it appear that the "scheme:" part of it is required at the beginning.  I have no strong opinion for whether the p4runtime_translation parameter needs a scheme included, or whether it should be assumed to always be "http://"

Reply:
I tried to emulate the notion of URL for a Protobuf `Any` message (https://developers.google.com/protocol-buffers/docs/proto3#any). I believe URI is more appropriate than URL for this, since for the `Any` message they seem to use the URL purely as an identifier.
We can probably be a bit liberal and use URI interchangeably with "URI suffix" in the spec. According to the same Wikipedia page, this seems to be a common practice.

Reply:
Thanks for checking.  Seems fine to me.